src/metabase/api/database.clj

```diff
@@ -261,22 +270,116 @@
 
 ;;; ------------------------------------------ GET /api/database/:id/fields ------------------------------------------
 
+(defn- call-on-first-row
+  "Calls `f` on the first row seen, acts as a no-op after"
+  [f]
+  (let [seen-first-row? (volatile! false)]
+    (fn [row]
+      (if @seen-first-row?
+        row
+        (do
+          (f row)
+          (vreset! seen-first-row? true)
+          row)))))
+
+(defn- json-piped-input-stream
+  "Ensures the response includes the JSON content type. This is needed as the body is an input stream and could be
+  interpretted as anything."
+  [f]
+  (rr/content-type {:status 200
+                    :body (rui/piped-input-stream f)}
+                   "application/json; charset=utf-8"))
+
+(defn- serialize-to-piped-input-stream
+  "Applies `transducer` to the reducible query results `reducible-q`. Serializes the results as JSON to a
+  PipedOutputStream. Returns the connected PipedInputStream as the body of a response map. Will call `on-first-row`
+  once the first row has be retrieved from `reducible-q` and calls `on-error` if an error occurs."
+  [on-first-row on-error transducer reducible-q]
+  (json-piped-input-stream
+   ;; This function will be invoked in a future and run in a different thread
+   (fn [^OutputStream output-stream]
+     (with-open [output-writer                 (OutputStreamWriter. ^OutputStream output-stream ^Charset StandardCharsets/UTF_8)
+                 buffered-writer               (BufferedWriter. output-writer)
+                 ^JsonGenerator json-generator (json/create-generator buffered-writer)]
+       (try
+         ;; We're always going to return a seq of objects, writing them in this way is the fastest way I've found
+         (.writeStartArray json-generator)
+         ;; This is a side-affecty way to run a reducible
+         (run! #(json-gen/encode-map % json-generator)
+               ;; This eduction will help us know that we've successfully executed a query and recevied a result from
+               ;; the result set. Up to this point, we have established a connection to the database but not run the
+               ;; query. We'll check for the first row first before running the `transducer` on the results
+               (eduction (map (call-on-first-row on-first-row)) transducer reducible-q))
```

Comment:
Why do you need eduction here? If I follow, the reducible is immediately (and only once) consumed by run! Can't you simply do:
```
(transduce (comp (map (call-on-first-row on-first-row)) transducer) reducible-q)
```

Reply:
Seems like an equal trade there. I guess the benefits of the above approach is that it doesn't wrap the reducible in the defrecord. I'd be curious what you think about this being nested in a with-open vs. the side-affecty transducer handling the opening/closing of the writer.

Reply:
If all things were equal, I'd stuff all state in the transducer, however your current approach is more robust in how it handles exceptions. If all state handling is in the transducer and there's an exception in the xf, you're left dangling.